### PR TITLE
feat: extend craft_application.git module

### PR DESCRIPTION
**Pull request by dariuszd21**
- [ ] Have you followed the guidelines for contributing?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `tox`?

-----

Add ability to manipulate repository and get additional data through `GitRepo` class:
- `set_remote_url` / `set_remote_push_url` alongside `get` counterparts (also `set_no_push` for disabling push for specific remotes)
- `is_commit()`, `is_short_commit` checkers
- use `craft.git` for subprocess git operation if available
- `get_last_commit()` to retrieve last commit on the current `HEAD`
- `fetch()` API to retrieve remote references

Linter issues fixed in #580 

(CRAFT-3708)

### CODE REVIEW

File: craft_application/git/_git_repo.py [hunk 1]
```diff
@@ -286,6 +338,119 @@ def rename_remote(
                 f"cannot rename '{remote_name}' to '{new_remote_name}'"
             ) from error
 
+    def show_remote(
+        self,
+        *,
+        remote_name_or_url: str,
+        do_not_query_remotes: bool = False,
+        disable_prompting: bool = True,
+    ) -> str:
```
Comment (dariuszd21) [inline craft_application/git/_git_repo.py:347]: Depending on the overflow discussion this may not end up being in the API at all

**Comment by tigarmo** [comment]
@dariuszd21 can you point out the code that hasn't been reviewed yet?

**Comment by dariuszd21** [comment]
> @dariuszd21 can you point out the code that hasn't been reviewed yet?

Sure, it will be all, but the first commit so:
`get_last_commit()` and `get_last_commit_on_branch_or_tag()` , changelog, integration tests for `set_url` and `set_push_url`, unit tests for `fetch` and few error cases for already tested APIs.
I'll mark snippets in the diff section

File: docs/reference/changelog.rst
Comment (dariuszd21) [inline docs/reference/changelog.rst:1]: All the changes are new here

File: tests/integration/git/test_git.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    """Check if remote push URL can be set using API."""
+    git_repo = GitRepo(empty_repository)
+    test_remote = "test-remote"
+    git_repo.add_remote(test_remote, hello_repository_lp_url)
+    assert git_repo.get_remote_url(remote_name=test_remote) == hello_repository_lp_url
+    assert git_repo.get_remote_push_url(remote_name=test_remote) is None
+
+    git_repo.set_no_push(test_remote)
+    assert git_repo.get_remote_url(remote_name=test_remote) == hello_repository_lp_url
+    assert git_repo.get_remote_push_url(remote_name=test_remote) == NO_PUSH_URL
```
Comment (dariuszd21) [inline tests/integration/git/test_git.py:97]: New integration test cases

File: craft_application/git/_git_repo.py [hunk 3]
```diff
@@ -196,6 +232,68 @@ def commit(self, message: str = "auto commit") -> str:
...
+            raise GitError("could not retrieve last commit") from error
+        else:
+            commit_message = cast(
+                str,
+                last_commit.message,  # pyright: ignore[reportAttributeAccessIssue,reportUnknownMemberType]
+            )
+            return Commit(
+                sha=str(last_commit.id),
+                message=commit_message,
+            )
```
Comment (dariuszd21) [inline craft_application/git/_git_repo.py:249]: Wrapped in the try except with additional error handling

File: craft_application/git/_git_repo.py [hunk 4]
```diff
@@ -196,6 +232,68 @@ def commit(self, message: str = "auto commit") -> str:
...
+            ) from error
+        else:
+            commit_message = cast(
+                str,
+                commit_obj.message,  # pyright: ignore[reportOptionalMemberAccess,reportAttributeAccessIssue,reportUnknownMemberType]
+            )
+            return Commit(
+                sha=commit_sha,
+                message=commit_message,
+            )
```
Comment (dariuszd21) [inline craft_application/git/_git_repo.py:295]: New functionality

File: craft_application/git/_git_repo.py [hunk 5]
```diff
@@ -382,6 +555,72 @@ def push_url(  # noqa: PLR0912 (too-many-branches)
             f"for the git repository in {str(self.path)!r}."
         )
 
+    def fetch(
+        self,
+        *,
+        remote: str,
+        tags: bool = False,
+        ref: str | None = None,
+        depth: int | None = None,
```
Comment (dariuszd21) [inline craft_application/git/_git_repo.py:564]: New arguments + handling them

File: tests/unit/git/test_git.py [hunk 6]
```diff
@@ -905,6 +1021,168 @@ def test_check_git_repo_for_remote_build_shallow(empty_working_directory):
...
+
+@pytest.mark.parametrize(
+    "tags", [True, False], ids=lambda x: "tags" if x else "no_tags"
+)
+@pytest.mark.parametrize(
+    "depth", [None, 1], ids=lambda x: "with_{x}_depth" if x else "no_depth"
+)
+@pytest.mark.parametrize(
+    "ref", [None, "aaaaaaa"], ids=lambda x: f"with_ref_{x}" if x else "no_ref_specified"
+)
```
Comment (dariuszd21) [inline tests/unit/git/test_git.py:1055]: Test case extended to accommodate new arguments

File: tests/unit/git/test_git.py [hunk 7]
```diff
@@ -985,3 +1263,83 @@ def test_describing_fallback_to_commit_for_unannotated_tags(
...
+) -> None:
+    git_repo = GitRepo(repository_with_commit.repository_path)
+    branch_or_tag = "test"
+    last_commit = git_repo.get_last_commit()
+    fake_process.register(
+        [expected_git_binary, "rev-list", "-n", "1", branch_or_tag],
+        stdout=last_commit.sha,
+    )
+    last_commit_on_branch = git_repo.get_last_commit_on_branch_or_tag(branch_or_tag)
+    assert last_commit_on_branch == last_commit
```
Comment (dariuszd21) [inline tests/unit/git/test_git.py:1345]: New tests

File: tests/unit/git/test_git.py [hunk 8]
```diff
@@ -905,6 +1021,168 @@ def test_check_git_repo_for_remote_build_shallow(empty_working_directory):
...
+    expected_git_binary: str,
+) -> None:
+    fake_process.register(
+        [expected_git_binary, "branch", "--remotes", "--contains", "fake-commit-sha"],
+        returncode=1,
+    )
+    git_repo = GitRepo(empty_repository)
+    with pytest.raises(GitError) as git_error:
+        git_repo.remote_contains(remote="fake-remote", commit_sha="fake-commit-sha")
+    assert git_error.value.details == "incorrect commit provided, cannot check"
```
Comment (dariuszd21) [inline tests/unit/git/test_git.py:1183]: New test

File: tests/unit/git/test_git.py [hunk 9]
```diff
@@ -905,6 +1021,168 @@ def test_check_git_repo_for_remote_build_shallow(empty_working_directory):
...
+    git_repo = GitRepo(empty_repository)
+    remote = "test-remote"
+    git_repo.add_remote(remote, "https://non-existing-repo.localhost")
+    cmd = [expected_git_binary, "fetch"]
+    cmd.append(remote)
+    fake_process.register(cmd, returncode=1)
+    with pytest.raises(GitError) as git_error:
+        git_repo.fetch(remote=remote)
+    assert git_error.value.details == f"cannot fetch remote: {remote!r}"
+
```
Comment (dariuszd21) [inline tests/unit/git/test_git.py:1094]: New test

File: tests/unit/git/test_git.py [hunk 10]
```diff
@@ -865,6 +874,113 @@ def test_check_git_repo_rename_remote_pygit_error_is_wrapped(
...
+def test_check_git_repo_set_no_push(mocker, empty_working_directory):
+    """Check if url is returned correctly using the get_url API."""
+    repo = GitRepo(empty_working_directory)
+    test_remote = "test-remote"
+
+    mocked_remotes = mocker.patch.object(repo._repo, "remotes")
+    repo.set_no_push(test_remote)
+
+    mocked_remotes.set_push_url.assert_called_with(test_remote, NO_PUSH_URL)
```
Comment (dariuszd21) [inline tests/unit/git/test_git.py:968]: New tests to increase coverage

**Review by dariuszd21 — COMMENTED** [review]
I marked where is a code that wasn't already reviewed for clarity

File: docs/reference/changelog.rst
Comment (tigarmo) [reply docs/reference/changelog.rst:1]: @medubelko can you take a look at these changelog entries please?

File: docs/reference/changelog.rst [hunk 11]
```diff
@@ -4,6 +4,34 @@
...
+4.6.0 (YYYY-MMM-DD)
+-------------------
+
+Git
+===
+
+- Extend ``craft_application.git`` module with the following APIs:
+
+  - Add ``is_commit(ref)`` and ``is_short_commit(ref)`` helpers for checking if
+    given ref is a valid commit hash.
```
Comment (tigarmo) [inline docs/reference/changelog.rst:16]:
```suggestion
    a given ref is a valid commit hash.
```

File: docs/reference/changelog.rst [hunk 12]
```diff
@@ -4,6 +4,34 @@
 Changelog
 *********
 
+4.6.0 (YYYY-MMM-DD)
+-------------------
+
+Git
+===
+
+- Extend ``craft_application.git`` module with the following APIs:
```
Comment (tigarmo) [inline docs/reference/changelog.rst:13]:
```suggestion
- Extend the ``craft_application.git`` module with the following APIs:
```

File: docs/reference/changelog.rst [hunk 13]
```diff
@@ -4,6 +4,34 @@
...
+
+  - Add ``is_commit(ref)`` and ``is_short_commit(ref)`` helpers for checking if
+    given ref is a valid commit hash.
+  - Add ``Commit`` model to represent ``get_last_commit()`` result.
+
+- Extend ``GitRepo`` class with additional methods:
```
Comment (tigarmo) [inline docs/reference/changelog.rst:19]:
```suggestion
- Extend the ``GitRepo`` class with additional methods:
```

File: docs/reference/changelog.rst [hunk 14]
```diff
@@ -4,6 +4,34 @@
...
+-------------------
+
+Git
+===
+
+- Extend ``craft_application.git`` module with the following APIs:
+
+  - Add ``is_commit(ref)`` and ``is_short_commit(ref)`` helpers for checking if
+    given ref is a valid commit hash.
+  - Add ``Commit`` model to represent ``get_last_commit()`` result.
```
Comment (tigarmo) [inline docs/reference/changelog.rst:17]:
```suggestion
  - Add a ``Commit`` model to represent the result of ``get_last_commit()``.
```

**Review by tigarmo — COMMENTED** [review]
thanks for pointing out the new changes!

File: craft_application/git/_git_repo.py
Comment (lengau) [inline craft_application/git/_git_repo.py:1]: Please add unit tests to cover the missing branches in this file.

File: craft_application/git/_models.py [hunk 15]
```diff
@@ -30,3 +31,16 @@ class GitType(Enum):
     INVALID = 0
     NORMAL = 1
     SHALLOW = 2
+
+
+@dataclass
+class Commit:
+    """Model representing commit."""
```
Comment (lengau) [inline craft_application/git/_models.py:38]:
```suggestion
    """Model representing a commit."""
```

File: craft_application/git/_git_repo.py [hunk 16]
```diff
@@ -482,3 +721,8 @@ def clone_repository(
                 f"cannot clone repository: {url} to {str(path)!r}"
             ) from error
         return cls(path)
+
+    @classmethod
+    def git_binary(cls) -> str:
```
Comment (lengau) [inline craft_application/git/_git_repo.py:726]:
```suggestion
    def get_git_command(cls) -> str:
```

File: craft_application/git/_git_repo.py [hunk 17]
```diff
@@ -120,6 +138,24 @@ def parse_describe(describe_str: str) -> str:
     return f"{version}.post{distance}+git{commit[1:]}"
 
 
+@lru_cache(maxsize=1)
+def _get_git_binary_name() -> str:
```
Comment (lengau) [inline craft_application/git/_git_repo.py:142]: What's the benefit of putting this in a protected function that just gets called from that classmethod?

Comment (dariuszd21) [reply craft_application/git/_git_repo.py:142]: Probably there is no benefit from doing that, I'll cache `classmethod` directly

Comment (dariuszd21) [reply craft_application/git/_git_repo.py]: Changed and tested in https://github.com/canonical/craft-application/pull/576/commits/ee680832348052d0af9df8a4d9c2094129957d6e

File: docs/reference/changelog.rst [hunk 18]
```diff
@@ -4,6 +4,34 @@
...
+
+- Extend the ``GitRepo`` class with additional methods:
+
+  - Add ``set_remote_url()`` and ``set_remote_push_url()`` alongside their get
+    counterparts.
+  - Add ``set_no_push()`` to explicitly disable ``push`` for specific remotes.
+  - Add ``get_last_commit()`` to retrieve last commit hash and message.
+  - Add ``get_last_commit_on_branch_or_tag()`` to retrieve last commit
+    associated with the given ref.
+  - Add ``fetch()`` method for retrieving remote objects.
```
Comment (medubelko) [inline docs/reference/changelog.rst:27]:
```suggestion
  - Add ``set_remote_url()`` and ``set_remote_push_url()`` methods and their
    getter counterparts.
  - Add ``set_no_push()`` method, which explicitly disables ``push`` for
    specific remotes.
  - Add ``get_last_commit()`` method, which retrieves the last commit hash and
    message.
  - Add ``get_last_commit_on_branch_or_tag()`` method, which retrieves the last
    commit associated with a given ref.
  - Add ``fetch()`` method, which retrieves remote objects.
```

File: docs/reference/changelog.rst [hunk 19]
```diff
@@ -4,6 +4,34 @@
...
+
+- Use ``craft.git`` for git-related operations done via ``subprocess`` in
```
Comment (medubelko) [inline docs/reference/changelog.rst:29]:
```suggestion
- Use ``craft.git`` for Git-related operations run with ``subprocess`` in
```

**Review by medubelko — CHANGES_REQUESTED** [review]
Some minor suggestions.

**Comment by dariuszd21** [comment]
> Some minor suggestions.

Thank you, applied you suggestions !

**Review by medubelko — APPROVED** [review]

**Pull request merged**